Add new subcommand "exec"

viktordick opened the pull request:
This is a generic subcommand that simplifies several git operations that
are automatically coupled with a playback.

Implements #70 
Existing tests run, but I need to actually test the functionality as well as the simplified versions of other commands.

viktordick commented:
@jan-jockusch Remembered again why I switched away from `exec`. `exec` is a builtint in Python 2, so the line

```
from .commands.exec import Exec
```

was invalid syntax. Of course it is not required that the file where the subcommand is defined has the same name, only lower case, but this is what I used up until now. I will change the file name to `execute.py`

viktordick commented:
Ready for review

gandie commented on perfact/zodbsync/subcommand.py:
```diff
@@ -113,6 +113,64 @@ def check_repo(self):
+                func(self, *args, **kwargs)
+                # Are these all?
+                for fname in ['rebase-merge', 'rebase-apply',
+                              'CHERRY_PICK_HEAD', 'MERGE_HEAD', 'REVERT_HEAD']:
```
These rascals should live in a named list, making this check more understandable. Something like `git_state_indicators`, i'm not sure. This list should live in a class or even module variable to allow monkey patching.

viktordick replied:
noted, good idea

gandie commented on perfact/zodbsync/subcommand.py:
```diff
@@ -113,6 +113,64 @@ def check_repo(self):
+                paths = sorted({
+                    filename[len(self.sync.site):].rsplit('/', 1)[0]
+                    for filename in files
+                })
```
Wow, nice combos! Computing what these lines do make my nose bleed. Some comments would be helpful, especially `filename[len(self.sync.site):].rsplit('/', 1)[0]` is hard to read ( on sunday morning without coffee ).

jan-jockusch replied:
I concur. On first read, I don't know jack what's going on here.

gandie reviewed (changes requested):
Nice one! I'd appreciate some minor code changes, but most importantly one test fails on my machine in the `py3` and `py2-zope5` environment:

```
________________________________________ TestSync.test_watch_move ________________________________________

self = <perfact.zodbsync.tests.test_sync.TestSync instance at 0x7f79b0291460>
conn = <perfact.zodbsync.helpers.Namespace instance at 0x7f79b0216230>

    def test_watch_move(self, conn):
        """
        Create a Page Template, record it using the watcher, rename it and make
        sure the watcher notices. Then add a second one and do a
        three-way-rename in one transaction, making sure the watcher keeps
        track.
        """
        watcher = self.runner('watch')
        watcher.setup()
        root = self.repo.path + '/__root__/'
        src = '/__source-utf8__.html'
        app = conn.app

        add = app.manage_addProduct['PageTemplates'].manage_addPageTemplate
        rename = app.manage_renameObject

        with conn.tm:
            add(id='test1', text='test1')
        watcher.step()
>       assert os.path.isdir(root + 'test1')
E       AssertionError: assert False
E        +  where False = <function isdir at 0x7f79bd098d50>(('/tmp/tmpT7EUcl/__root__/' + 'test1'))
E        +    where <function isdir at 0x7f79bd098d50> = <module 'posixpath' from '/home/lars/Dev/zodbsync/.tox/py2-zope5/lib/python2.7/posixpath.pyc'>.isdir
E        +      where <module 'posixpath' from '/home/lars/Dev/zodbsync/.tox/py2-zope5/lib/python2.7/posixpath.pyc'> = os.path

perfact/zodbsync/tests/test_sync.py:363: AssertionError
```

Not sure what's going on here, `py2-zope2` environment seems to be cool with it:

```
________________________________________________ summary _________________________________________________
ERROR:   py3: commands failed
ERROR:   py2-zope5: commands failed
  py2-zope2: commands succeeded
```

This could be a problem on my machine, on another machine running a higher tox version tests are running fine. Lets talk about this and investigate / compare setup, I'd like to see this working on all machines.

viktordick commented:
Weird. Do you have some kind of a cron job that cleans up temporary files maybe?
Also, I tested it myself a little bit more and added one commit. And I noticed that the rollback does not play nice when switching branches (if my command contains a checkout onto another branch, it resets that other branch to my original HEAD). Not sure how to address this, maybe we need to remember the state of each branch and compare them. Maybe only checking the checked-out branch before and after is enough.

Marking this as draft for now.

jan-jockusch commented on README.md:
```diff
@@ -124,10 +129,27 @@ marked as `unsupported`, which are ignored if found in the ZODB. If only a
+
+This allows commands like the following:
+
+    zodbsync exec "git cherry-pick COMMIT"
```
I'm not sure whether this is possible, but it would be nice to be able to drop the quotes, and let "zodbsync exec" behave like other Unix exec wrappers, e.g. "timeout", "watch", and "flock". The examples would then read:

    zodbsync exec git cherry-pick COMMIT
    and so on.

viktordick replied:
I do not think this is possible with `argparse`

jan-jockusch replied:
Alright, it was just a thought.

jan-jockusch reviewed (commented):
Nothing really new to contribute. Just an idea of wrapping without quotes.

viktordick commented:
Extended the rollback mechanism to reset all branches. Existing tests run, but need to add tests for this.

viktordick commented:
I now also had the watcher tests misbehaving sometimes, although after a few runs where I also tried to set a debugger at the corresponding points, I can now no longer reproduce it.

@gandie If you can reproduce it consistently, could you please check if it occurs only on this branch or already on master? I don't think it has something to do with these changes (it occurs in the watcher tests), so we should tackle it separately.

Also, I added a test for checking out another branch and fixed the things you mentioned. This is ready for re-review.

viktordick commented:
Added both of my faithful reviewers, just to allow the first one that finds the time to give a review.

jan-jockusch reviewed (approved):
LGTM!
